File: src/cpp/config_store.cpp
#include "config_store.h"

namespace fastly::config_store {

ConfigStore ConfigStore::open(std::string_view name) {
  return ConfigStore(
      fastly::sys::config_store::m_static_config_store_config_store_open(
          static_cast<std::string>(name)));
}

std::optional<std::string> ConfigStore::get(std::string_view key) {
  std::string out;
  auto some{this->cs->get(static_cast<std::string>(key), out)};
  if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>{std::move(out)};
  }
}

bool ConfigStore::contains(std::string_view key) {
  return this->cs->contains(static_cast<std::string>(key));
}

} // namespace fastly::config_store

File: src/config_store.rs
use std::pin::Pin;

use cxx::CxxString;

pub struct ConfigStore(pub(crate) fastly::ConfigStore);

pub fn m_static_config_store_config_store_open(name: &CxxString) -> Box<ConfigStore> {
    Box::new(ConfigStore(fastly::ConfigStore::open(
        name.to_string_lossy().as_ref(),
    )))
}

impl ConfigStore {
    pub fn get(&self, key: &CxxString, out: Pin<&mut CxxString>) -> bool {
        self.0
            .get(key.to_string_lossy().as_ref())
            .map(|val| out.push_str(val.as_ref()))
            .is_some()
    }

    pub fn contains(&self, key: &CxxString) -> bool {
        self.0.contains(key.to_string_lossy().as_ref())
    }
}
